pkg/controllers/certificate/cert_rotation_controller.go

```diff
@@ -0,0 +1,315 @@
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *CertRotationController) SetupWithManager(mgr controllerruntime.Manager) error {
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&clusterv1alpha1.Cluster{}, builder.WithPredicates(c.PredicateFunc)).
+		WithEventFilter(predicate.GenerationChangedPredicate{}).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiterflag.DefaultControllerRateLimiter(c.RatelimiterOptions),
+		}).
+		Complete(c)
+}
+
+func (c *CertRotationController) syncCertRotation(secret *corev1.Secret) error {
+	karmadaKubeconfig, err := getKubeconfigFromSecret(secret)
+	if err != nil {
+		return err
+	}
+
+	clusterName := karmadaKubeconfig.Contexts[karmadaKubeconfig.CurrentContext].AuthInfo
+	if clusterName == "" {
+		return fmt.Errorf("failed to get cluster name, the current context is %s", karmadaKubeconfig.CurrentContext)
+	}
+
+	oldCertData := karmadaKubeconfig.AuthInfos[clusterName].ClientCertificateData
+
+	shouldRotate, err := c.shouldRotateCert(oldCertData)
+	if err != nil {
+		return err
+	}
+	if !shouldRotate {
+		return nil
+	}
+
+	oldCert, err := certutil.ParseCertsPEM(oldCertData)
+	if err != nil {
+		return fmt.Errorf("unable to parse old certificate: %v", err)
+	}
+
+	// create a new private key
+	keyData, err := keyutil.MakeEllipticPrivateKeyPEM()
+	if err != nil {
+		return err
+	}
+
+	privateKey, err := keyutil.ParsePrivateKeyPEM(keyData)
+	if err != nil {
+		return fmt.Errorf("invalid private key for certificate request: %v", err)
+	}
+
+	csr, err := c.createCSRInControlPlane(clusterName, privateKey, oldCert)
+	if err != nil {
+		return fmt.Errorf("failed to create csr in control plane, err is: %v", err)
+	}
+
+	var newCertData []byte
+	klog.V(1).Infof("Waiting for the client certificate to be issued")
+	err = wait.Poll(1*time.Second, 5*time.Minute, func() (done bool, err error) {
+		csr, err := c.KubeClient.CertificatesV1().CertificateSigningRequests().Get(context.TODO(), csr, metav1.GetOptions{})
+		if err != nil {
+			return false, fmt.Errorf("failed to get the cluster csr %s. err: %v", clusterName, err)
+		}
+
+		if csr.Status.Certificate != nil {
+			klog.V(1).Infof("Signing certificate successfully")
+			newCertData = csr.Status.Certificate
+			return true, nil
+		}
+
+		klog.V(1).Infof("Waiting for the client certificate to be issued")
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+
+	karmadaKubeconfig.AuthInfos[clusterName].ClientCertificateData = newCertData
+	karmadaKubeconfig.AuthInfos[clusterName].ClientKeyData = keyData
+
+	karmadaKubeconfigBytes, err := clientcmd.Write(*karmadaKubeconfig)
+	if err != nil {
+		return fmt.Errorf("failed to serialize karmada-agent kubeConfig. %v", err)
+	}
+
+	secret.Data["karmada-kubeconfig"] = karmadaKubeconfigBytes
+	// Update the karmada-kubeconfig secret in the member cluster.
+	if err := util.CreateOrUpdateSecret(c.ClusterClient.KubeClient, secret); err != nil {
+		return err
+	}
```

Comment:
Why do we need to try `create` the secret? We just get it from APIserver, what we need is to make an update, isn't it?

Reply:
The function is  for convenience mainly. The main possible concern is the performance of using get to determine presence and using create to determine presence.